sklearn/tree/_classes.py

```diff
@@ -336,13 +339,45 @@ def fit(self, X, y, sample_weight=None, check_input=True):
         SPLITTERS = SPARSE_SPLITTERS if issparse(X) else DENSE_SPLITTERS
 
         splitter = self.splitter
+        if self.monotonic_cst is None:
+            monotonic_cst = np.zeros(shape=X.shape[1], dtype=np.int8)
+        else:
+            if self.n_outputs_ > 1:
+                raise ValueError(
+                    "Monotonic constraints are not supported with multiple outputs."
+                )
+            # Applying element-wise logical conjunction
+            # for monotonic constraints' support.
+            monotonic_cst = np.asarray(self.monotonic_cst, dtype=np.int8)
+            unsatisfied_constraints_conditions = (
+                (monotonic_cst != -1) * (monotonic_cst != 0) * (monotonic_cst != 1)
+            )
+            if np.any(unsatisfied_constraints_conditions):
+                raise ValueError(
+                    "monotonic_cst must be None or an array-like of -1, 0 or 1."
+                )
+            if is_classifier(self):
+                if self.n_classes_[0] > 2:
+                    raise ValueError(
+                        "Monotonic constraints are not supported with multiclass "
+                        "classification"
+                    )
+                # Imposing the constraint on the probability of the positive class
+                monotonic_cst *= -1
+
+        if monotonic_cst.shape[0] != X.shape[1]:
+            raise ValueError(
+                "monotonic_cst has shape {} but the input data "
+                "X has {} features.".format(monotonic_cst.shape[0], X.shape[1])
+            )
```

Comment:
I think this validation logic should be factorized in a dedicated `sklearn.utils._validation.check_monotonic_cst` and be reused for `HistGradientBoosting*` as well.